src/Lykke.blue.Service.ReferralLinks.AzureRepositories/ReferralLink/ReferralLinkRepository.cs

```diff
@@ -1,24 +1,26 @@
 ﻿using AutoMapper;
 using AzureStorage;
-using Lykke.blue.Service.ReferralLinks.AzureRepositories.DTOs;
 using Lykke.blue.Service.ReferralLinks.Core.Domain.ReferralLink;
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace Lykke.blue.Service.ReferralLinks.AzureRepositories.ReferralLink
 {
     public class ReferralLinkRepository : IReferralLinkRepository
     {
         private readonly INoSQLTableStorage<ReferralLinkEntity> _referralLinkTable;
+        private static readonly SemaphoreSlim SemaphoreSlim = new SemaphoreSlim(1, 1);
+        private const string ReflinkPartitionKey = "ReferallLink";
```

Comment:
How many record we can have in this ReflinkPartitionKey?
if there are many entries in one partition, then Azure will not work efficiently and will not be able to optimize

Reply:
Using a single value for the PartitionKey is mainly dictated on how we retrieve/read data. Here are the reasons/benefits of having a constant partition key:
 - Referral links are claimed/queried by the client app by refLink Id which is equal to the RowKey Id in db. In certain scenarios, we don't have any other input parameters from the client app, and thats why having a constant partition key allows us to us Point queries (speechifying both partition key & raw key) which are the fastest possible as far as Table storage is concerned.(as opposed to searching by raw key only)
- We have a few queries  (GetExpiredGiftCoinLinks(), GetReferralLinksBySenderId(), GetReferalLinkByUrl()), which are based entirely on non-key entity fields (senderId, DateTime, Type, etc). As the PartitionKey is constant, we can include it in the queries, thus making the quires run faster. Otherwise, having multiple partition keys, will force us to search by non key fields only (because we dont know the exact partition key)
- We use batch transactions. We have these new endpoints allowing for a mass generation for ref links where we insert all new ref links in db in a single batch operation. Batch operations are only possible if all entities are being inserted into the same partition. (Actually, we can use the sender client id as a partition key for the batch inserted entities, which will allow the batch to go through, but all the other issues with having multiple partition keys will remain)
- We already have some ref links generated on PROD with ReflinkPartitionKey = "ReferallLink". They are not that many but we will have to discard them if we change the code to use unique/multiple partitions.

I did some tests on the performance of having a single partition key. I have inserted 100000 rows in the table, all belonging to to the same partition key, then tested the read/write performance:

- look up (read) a single raw from db by PartitionKey and RawKey:  ~1 sec
- insert a single raw in db: ~1 sec
- look up (read) a single raw from db by non key columns (Type, ExpDate, etc): ~40s

I am not sure if having multiple partitions will speed up the query by non key columns much, as it will be a table scan because we cant specify the exact partition key (we wouldn't know it).

Nevertheless, we have two options for using unique/multiple partition keys, if you think it would be better in the long term:
- We can set the PartitionKey to be equal to the RawKey (which is a GUID). This effectively will give us unique partition key for every table entry. The cons are that we cant use batch inserts (we will insert ref links one by one, I have inserted 1000 ref links one by one for 1 min, which is not bad in fact). Also, the non-key column queries will somehow slower, but I dont know whether it will be more or less than 40s.
- The other option for a PartitionKey is the SenderClientId (the client who generated the ref links). This way all ref links generated by a single client will belong to the same partition - this will give us multiple partition keys) The problem here is that the look up by ref link Id, will only be by RawKey, as the partition key will vary. So it will be somehow slower (not much I guess) and in theory we may have duplicates , if the same GUID is generated twice (highly unlikly though) - i.e. the service runs on different machines etc.